Improve internal retry scheme

louis-z opened the pull request:
- Disable internal retry scheme for non-idempotent HTTP requests
- Convert `FunctionRetryExtensionsTests.Retry` from recursive to iterative
- Log debug info about internal retries during tests

louis-z committed:
For requests other than GETs, disable internal retry scheme

Plus, log info during tests about internal retries.

louis-z commented:
Some tests are still randomly failing... Let's see where this PR takes us.

meziantou commented on NGitLab/Impl/HttpRequestor.GitLabRequest.cs:
```diff
@@ -58,6 +58,11 @@ public GitLabRequest(Uri url, MethodType method, object data, string apiToken, s
 
             public WebResponse GetResponse(RequestOptions options)
             {
+                // For any requests that are POTENTIALLY NOT idempotent, send only once
```
I think we should update the RequestOptions with a new method 

````
public virtual bool ShouldRetry(Exception ex, WebResponse response, int retryNumber)
````

We can change the default ShouldRetry implementation to skip non-GET methods. But we still allow people to retry POST method if they want to.

louis-z replied:
I like the idea of giving the user the possibility to choose the old behavior. However, I'm not sure that adding a new ShouldRetry (taking in an additional WebResponse) is the best approach, for at least a few reasons:
- The Retry\<T> method is generic and therefore does not know about WebResponse. We could probably try to cast T to WebResponse and work with that, if it weren't for the other reasons.
- In the said method, the WebResponse is actually returned by calling action(), which actually triggers the retry when it throws. At that point there is no returned WebResponse to pass in to ShouldRetry.
- I think that a simpler approach is just to offer the user a bool to indicate they want the old behavior.

louis-z replied:
Ok, so I revisited the code as follows:

 - GitLabException now stores the HTTP request method, if any, that triggered the exception
- RequestOptions.ShouldRetry was adapted so that
  - HTTP request methods that are NOT safe will not be retried
  - We use the same Transient HTTP StatusCodes as Polly, i.e. >= 500 or == 408

louis-z committed:
Add RequestOptions.RetrySafeRequestsOnly + consider GET and HEAD as safe

louis-z committed:
Store HTTP method in GitLabException and adapt ShouldRetry

- GitLabException provides the HTTP request method that triggered the exception, if any
- RequestOptions.ShouldRetry was adapted so that
  - HTTP request methods that are NOT safe will not be retried
  - We use the same Transient HTTP StatusCodes as Polly, i.e. >= 500 or == 408

louis-z committed:
fixup

meziantou commented on NGitLab/RequestOptions.cs:
```diff
@@ -39,7 +39,24 @@ public RequestOptions(int retryCount, TimeSpan retryInterval, bool isIncremental
+            // If there is no HTTP request method specified, carry on to the next condition below.
+            if (gitLabException.MethodType.HasValue &&
+                gitLabException.MethodType != Impl.MethodType.Get &&
+                gitLabException.MethodType != Impl.MethodType.Head)
```
```suggestion
                gitLabException.MethodType != Impl.MethodType.Head &&
                gitLabException.MethodType != Impl.MethodType.Options)
```

louis-z committed:
Convert FunctionRetryExtensionsTests.Retry from recursive to iterative

louis-z committed:
Consider the "OPTIONS" HTTP request method safe to retry

louis-z committed:
Fix error message

meziantou commented on NGitLab/Extensions/FunctionRetryExtensions.cs:
```diff
@@ -8,22 +8,32 @@ namespace NGitLab.Extensions
+                {
+                    return action();
+                }
+                catch (Exception ex) when (predicate(ex, retriesLeft))
```
We should force the handling of `retriesLeft`

```suggestion
                catch (Exception ex) when (retriesLeft > 0 && predicate(ex, retriesLeft))
```

louis-z replied:
The `predicate` should handle it, as before. Otherwise what's the point in passing in `retriesLeft` to the predicate?

If a user overrides the default predicate and forgets to take `retriesLeft` into account, well I say tough luck. 😁

meziantou replied:
> Otherwise what's the point in passing in retriesLeft to the predicate?

logging/informative, but I think we should enforce it to avoid bad implementations

meziantou commented on NGitLab/Extensions/FunctionRetryExtensions.cs:
```diff
@@ -8,22 +8,32 @@ namespace NGitLab.Extensions
+                    Thread.Sleep(waitTime);
+
+                    if (useExponentialBackoff)
+                        waitTime = waitTime.Add(waitTime);
```
```suggestion
                    {
                        waitTime = waitTime.Add(waitTime);
                    }                        
```

louis-z committed:
Add braces

louis-z committed:
Consider retriesLeft in Retry method
